Doc clarification regarding Logstash and OpenShift SCC in ECK 3.4.0+ (#9552)
elastic/cloud-on-k8s: docs/release-notes/known-issues.md

To check whether your namespace enforces restricted PSA, run:

```bash
kubectl get namespace <your-namespace> -o jsonpath='{.metadata.labels}'

# or on OpenShift:
oc get namespace <your-namespace> -o jsonpath='{.metadata.labels}'
```

If you see `pod-security.kubernetes.io/enforce: restricted`, do **not** use this workaround — in restricted-PSA namespaces, `seccompProfile` is required and must not be omitted, and `fsGroup` must fall within the namespace's allocated `supplemental-groups` range rather than a fixed value like `1000`.

For namespaces using `anyuid`-style SCC without restricted PSA enforcement, override the Logstash pod security context as follows to prevent ECK from injecting its defaults:

```yaml
spec:
  podTemplate:
    spec:
      securityContext:
        fsGroup: 1000
        # seccompProfile intentionally omitted — not permitted by anyuid SCC
```

This override can be removed once you upgrade to ECK 3.5.0 or later when available, which includes the fix for [#9550](https://github.com/elastic/cloud-on-k8s/issues/9550).

> :::

Reviewer: Is this known issue exactly the same as the one reported in 3.4.1? If so, and considering how long the block is, I'd strongly suggest to put the text into a snippet and import it in both versions, so we can have a single source of truth.

For the content of the workaround we need ensure @pebrc and the team validates it, as I've seen an outdated comment where he mentions that the proposed workaround wasn't completely valid. Just in case that's still applicable.

Reviewer 2: Hi @eedugon 👋 the workaround has been validated on OpenShift with the v3.4.0 operator.

One important clarification for the docs though, @kunisen: the podTemplate.spec.securityContext override must be applied to existing Logstash CRs before upgrading to ECK 3.4.0. If the upgrade happens first and the Logstash pods enter the SCC rejection loop, updating the CR afterwards won't help - the controller detects pending pod creations and re-queues without patching the StatefulSet spec, so the fix never takes effect. For the latter, the only recovery is to delete and re-create the Logstash CR with the security context as described in place.

kubectl get namespace <your-namespace> -o jsonpath='{.metadata.labels}'

# or on OpenShift:
oc get namespace <your-namespace> -o jsonpath='{.metadata.labels}'
```

If the output contains `pod-security.kubernetes.io/enforce: restricted`, do **not** apply this workaround: restricted PSA requires `seccompProfile` and expects `fsGroup` within the namespace's allocated `supplemental-groups` range, not a fixed `1000`.

Otherwise, set the security context on the Logstash resource so ECK skips its own injection:

```yaml
spec:
  podTemplate:
    spec:
      securityContext:
        fsGroup: 1000
        # seccompProfile intentionally omitted — not permitted by anyuid SCC
```

**Important:** Apply this override to existing Logstash resources **before** upgrading. If you upgrade first and the pods enter the SCC rejection loop, editing the resource afterwards has no effect — the only recovery is to delete and re-create the Logstash resource with the override already in place.

You can remove this override after upgrading to ECK 3.5.0 or later (once available), which fixes [#9550](https://github.com/elastic/cloud-on-k8s/issues/9550).
::::

## 3.4.0 [elastic-cloud-kubernetes-340-known-issues]

> :::{dropdown} Logstash pods rejected by OpenShift SCC when using a non-default security context constraint

Reviewer: ```suggestion
::::{dropdown} Logstash pods rejected by OpenShift SCC when using a non-default security context constraint
```

Adding an extra `:` (I'm also adding an extra colon in the dropdown closing line)

**Workaround**

This workaround applies only to namespaces governed by an `anyuid`-style SCC (`RunAsAny` fsGroup, seccomp forbidden) that are **not** also enforcing [restricted Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/).

To check whether your namespace enforces restricted PSA, run:

```bash
kubectl get namespace <your-namespace> -o jsonpath='{.metadata.labels}'

# or on OpenShift:
oc get namespace <your-namespace> -o jsonpath='{.metadata.labels}'
```

If you see `pod-security.kubernetes.io/enforce: restricted`, do **not** use this workaround — in restricted-PSA namespaces, `seccompProfile` is required and must not be omitted, and `fsGroup` must fall within the namespace's allocated `supplemental-groups` range rather than a fixed value like `1000`.

For namespaces using `anyuid`-style SCC without restricted PSA enforcement, override the Logstash pod security context as follows to prevent ECK from injecting its defaults:

```yaml
spec:
  podTemplate:
    spec:
      securityContext:
        fsGroup: 1000
        # seccompProfile intentionally omitted — not permitted by anyuid SCC
```

> **Important:** Apply this override to existing Logstash resources **before** upgrading. If you upgrade first and the pods enter the SCC rejection loop, editing the resource afterwards has no effect — the only recovery is to delete and re-create the Logstash resource with the override already in place.

Reviewer: ```suggestion
:::{important}
Apply this override to existing Logstash resources before upgrading to ECK 3.4. If you are already running ECK 3.4 and the pods are in the SCC rejection loop, delete the Logstash resource and re-create it with the override described in this workaround.
:::
```

We should avoid paragraphs starting with **Important**, or **Note**, as we have admonitions for that purpose.

Converting to an admonition and rephrasing a bit. Note that **upgrade** is not the only way to reach this situation, a 3.4 installation could face this problem too. That's why I propose to change `if you upgrade first` to `If you are running ECK 3.4 already`.

You can remove this override after upgrading to ECK 3.5.0 or later (once available), which fixes [#9550](https://github.com/elastic/cloud-on-k8s/issues/9550).

> :::

Reviewer: ```suggestion
::::
```

Adding an extra `:` to be able to include the important admonition inside with the right element.

For more information, check this [Issue #9550](https://github.com/elastic/cloud-on-k8s/issues/9550).

**Workaround**

This workaround applies **only** to namespaces using an `anyuid`-style SCC (`RunAsAny` fsGroup, seccomp forbidden) that do **not** enforce [restricted Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/). Check your namespace's labels first:

```bash
kubectl get namespace <your-namespace> -o jsonpath='{.metadata.labels}'

# or on OpenShift:
oc get namespace <your-namespace> -o jsonpath='{.metadata.labels}'
```

If the output contains `pod-security.kubernetes.io/enforce: restricted`, do **not** apply this workaround: restricted PSA requires `seccompProfile` and expects `fsGroup` within the namespace's allocated `supplemental-groups` range, not a fixed `1000`.

Otherwise, set the security context on the Logstash resource so ECK skips its own injection:

```yaml
spec:
  podTemplate:
    spec:
      securityContext:
        fsGroup: 1000
        # seccompProfile intentionally omitted — not permitted by anyuid SCC
```

> **Important:** Apply this override to existing Logstash resources **before** upgrading. If you upgrade first and the pods enter the SCC rejection loop, editing the resource afterwards has no effect — the only recovery is to delete and re-create the Logstash resource with the override already in place.

Reviewer 2: this should be consistent right?

```suggestion
:::{important}
Apply this override to existing Logstash resources before upgrading to ECK 3.4. If you are already running ECK 3.4 and the pods are in the SCC rejection loop, delete the Logstash resource and re-create it with the override described in this workaround.
:::
```

## 3.4.1 [elastic-cloud-kubernetes-341-known-issues]

> :::{dropdown} Logstash pods rejected by OpenShift SCC when using a non-default security context constraint

Reviewer 3: **Rendering bug (confirmed in the docs preview):** the `:::{important}` block below is nested inside this `:::{dropdown}` using the same number of colons, so it doesn't nest. As a result the closing sentence "This override can be removed once you upgrade to ECK 3.5.0…" renders *outside* the dropdown as loose body text. The outer directive must use more colons than the inner one. Bump this opening fence and its matching closing fence (line 58) to 4 colons; leave the inner `:::{important}` at 3. The 3.4.0 copy (lines 62 & 98) needs the same fix.

```suggestion
::::{dropdown} Logstash pods rejected by OpenShift SCC when using a non-default security context constraint
```

> In ECK 3.4.x, the Logstash controller unconditionally injects `seccompProfile: RuntimeDefault` and `fsGroup: 1000` into the pod security context, ignoring the `--set-default-security-context=auto-detect` operator flag. On OpenShift, this flag should suppress the injection, as it does for all other ECK-managed workloads ({{es}}, {{product.kibana}}, APM Server). Clusters using a non-default SCC such as `anyuid` — which does not permit explicit seccomp configuration — will see Logstash pods rejected at admission after upgrading.

Reviewer 3: Tightening: `always` reads better than `unconditionally`, and splitting the final clause makes the cause/effect clearer.

```suggestion
In ECK 3.4.x, the Logstash controller always injects `seccompProfile: RuntimeDefault` and `fsGroup: 1000` into the pod security context, ignoring the `--set-default-security-context=auto-detect` flag. On OpenShift, this flag should suppress the injection — as it does for {{es}}, {{product.kibana}}, and APM Server. As a result, clusters using a non-default SCC such as `anyuid`, which forbids explicit seccomp settings, reject Logstash pods at admission after upgrading.
```

For more information, check this [Issue #9550](https://github.com/elastic/cloud-on-k8s/issues/9550).

**Workaround**

This workaround applies only to namespaces governed by an `anyuid`-style SCC (`RunAsAny` fsGroup, seccomp forbidden) that are **not** also enforcing [restricted Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/).

> To check whether your namespace enforces restricted PSA, run:

Reviewer 3: Fold the "To check … run:" lead-in into the sentence above — three lines become one.

```suggestion
This workaround applies **only** to namespaces using an `anyuid`-style SCC (`RunAsAny` fsGroup, seccomp forbidden) that do **not** enforce [restricted Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/). Check your namespace's labels first:
```

```bash
kubectl get namespace <your-namespace> -o jsonpath='{.metadata.labels}'

# or on OpenShift:
oc get namespace <your-namespace> -o jsonpath='{.metadata.labels}'
```

> If you see `pod-security.kubernetes.io/enforce: restricted`, do **not** use this workaround — in restricted-PSA namespaces, `seccompProfile` is required and must not be omitted, and `fsGroup` must fall within the namespace's allocated `supplemental-groups` range rather than a fixed value like `1000`.

Reviewer 3: Tighten.

```suggestion
If the output contains `pod-security.kubernetes.io/enforce: restricted`, do **not** apply this workaround: restricted PSA requires `seccompProfile` and expects `fsGroup` within the namespace's allocated `supplemental-groups` range, not a fixed `1000`.
```

> For namespaces using `anyuid`-style SCC without restricted PSA enforcement, override the Logstash pod security context as follows to prevent ECK from injecting its defaults:

Reviewer 3: Shorter lead-in to the YAML.

```suggestion
Otherwise, set the security context on the Logstash resource so ECK skips its own injection:
```

```yaml
spec:
  podTemplate:
    spec:
      securityContext:
        fsGroup: 1000
        # seccompProfile intentionally omitted — not permitted by anyuid SCC
```

:::{important}

> The `podTemplate.spec.securityContext` override must be applied to existing Logstash CRs before upgrading to ECK 3.4.0. If the upgrade happens first and the Logstash pods enter the SCC rejection loop, updating the CR afterwards won't help. For the latter, the only recovery is to delete and re-create the Logstash CR with the security context as described in place.

Reviewer 3: Tighten, drop the confusing "before upgrading to ECK 3.4.0" (this is the 3.4.1 section), and replace the awkward "as described in place". Also avoiding the `CR` abbreviation in favour of "Logstash resource".

```suggestion
Apply this override to existing Logstash resources **before** upgrading. If you upgrade first and the pods enter the SCC rejection loop, editing the resource afterwards has no effect — the only recovery is to delete and re-create the Logstash resource with the override already in place.
```

:::

> This override can be removed once you upgrade to ECK 3.5.0 or later when available, which includes the fix for [#9550](https://github.com/elastic/cloud-on-k8s/issues/9550).

Reviewer 3: Tighten.

```suggestion
You can remove this override after upgrading to ECK 3.5.0 or later (once available), which fixes [#9550](https://github.com/elastic/cloud-on-k8s/issues/9550).
```

> :::

Reviewer 3: This does not render correctly: 

<img width="885" height="1005" alt="Image" src="https://github.com/user-attachments/assets/395e797c-d027-49d0-8bbf-56e2e0186113" />

> :::

Reviewer 3: Close the dropdown with 4 colons to match the opening fence (see the nesting note above).

```suggestion
::::
```

## 3.4.0 [elastic-cloud-kubernetes-340-known-issues]

> :::{dropdown} Logstash pods rejected by OpenShift SCC when using a non-default security context constraint

Reviewer 3: This 3.4.0 block (lines 62–98) is byte-for-byte identical to the 3.4.1 one above, so every suggestion above applies here too. Rather than applying each fix twice, consider @eedugon's suggestion of extracting the block into a shared snippet and importing it into both sections — single source of truth, and the structural fence fix only has to be made once.

Reviewer 2: @kunisen any hard blocker we can't do that?

## 3.4.0 [elastic-cloud-kubernetes-340-known-issues]

:::{dropdown} Logstash pods rejected by OpenShift SCC when using a non-default security context constraint
In ECK 3.4.x, the Logstash controller unconditionally injects `seccompProfile: RuntimeDefault` and `fsGroup: 1000` into the pod security context, ignoring the `--set-default-security-context=auto-detect` operator flag. On OpenShift, this flag should suppress the injection, as it does for all other ECK-managed workloads ({{es}}, {{product.kibana}}, APM Server). Clusters using a non-default SCC such as `anyuid` — which does not permit explicit seccomp configuration — will see Logstash pods rejected at admission after upgrading.

For more information, check this [Issue #9550](https://github.com/elastic/cloud-on-k8s/issues/9550).

**Workaround**

Override the Logstash pod security context explicitly to prevent ECK from injecting its defaults. Keep `fsGroup: 1000` as Logstash requires it for volume writes, but omit `seccompProfile`:

```yaml
spec:
  podTemplate:
    spec:
      securityContext:
        fsGroup: 1000
        # seccompProfile intentionally omitted — not permitted by anyuid SCC

> ```

Reviewer 3: This workaround is not general. It's valid **only** in namespaces governed by an `anyuid`-style SCC (`RunAsAny` fsGroup, seccomp forbidden) that are **not** also enforcing restricted Pod Security Admission. 

Recommend: scope this explicitly to anyuid-type SCCs without restricted PSA, add a warning that in restricted-PSA namespaces `seccompProfile` is required (do not omit) and `fsGroup` must fall within the namespace's allocated `supplemental-groups` range (not `1000`). As written, we'd be publishing a workaround that is not generally applicable.

> This override can be removed once you upgrade to the ECK version that includes the fix for [#9550](https://github.com/elastic/cloud-on-k8s/issues/9550).

Reviewer 2: @eedugon should we refer to future versions?

Author: Thanks @pkoutsovasilis 
Also to @eedugon 

I found a better phrase to avoid potential confusion - `or later once available`, per https://www.elastic.co/docs/release-notes/cloud-on-k8s/known-issues

<img width="760" height="471" alt="image" src="https://github.com/user-attachments/assets/94d3d312-a563-425d-a225-460ee2d0b382" />

I will use the below phrase since https://github.com/elastic/cloud-on-k8s/pull/9551 seems to be included in 3.5.0:

```
Alternatively, upgrade to ECK 3.5.0 or later once available.
```

Author: @pkoutsovasilis @eedugon this is the updated wording - https://github.com/elastic/cloud-on-k8s/pull/9552/changes/7de745d3ed5a4de61585934c06e2e993b2010443. Thanks!

Reviewer: > @eedugon should we refer to future versions?

Interesting topic. The current proposal doesn't look bad, although in some way it doesn't feel ideal.

I'll check internally with the team to see if there's a better way of handling this type of statement.

Reviewer 2: @eedugon do you have any update on this one?

Reviewer: @pkoutsovasilis , we don't have a definitive solution or specific policy here.

If we are almost sure that our next minor release will solve the issue and that release is not far away in time then I'd say it's ok to use the approach shared by @kunisen's screenshot that is already in place in other parts of the release notes. For example:

> Alternatively, upgrade to ECK 3.5.0 or later once available.

And ideally, when 3.5.0 is shipped it would be awesome to update the sentence and convert it to something like:

> This issue is fixed in ECK 3.5.0 (or whatever way that flows properly in the narrative)

But if we are uncertain of when the issue will be fixed we should stick to the original proposal, with something like:

> This override can be removed once you upgrade to the ECK version that includes the fix for [#9550](https://github.com/elastic/cloud-on-k8s/issues/9550). (or any similar way, just implying that the issue will be fixed in a future release and the implications to versions facing the issue).

With this second approach I would consider important (not ideal) to update the known issue report as soon as the issue is really fixed, as that type of text, if it's outdated, looks worse than an outdated `upgrade to ECK 3.5.0 or later once available`.

In short: Better to not refer to future releases but we have to find a balance in terms of docs maintainability. So whatever you prefer and feel it's better.
